concurrency/encoding.go

```diff
@@ -3,14 +3,73 @@ package concurrency
 import (
 	"compress/gzip"
 	"errors"
+	"fmt"
 	"io"
 	"sync"
 
 	jsoniter "github.com/json-iterator/go"
 	yaml "gopkg.in/yaml.v3"
 )
 
-var gzipWPool, gzipRPool sync.Pool
+var (
+	gzipWPool, gzipRPool sync.Pool
+
+	// ErrExpectByteSlicePtr denotes that the assertion of a byte slice pointer failed
+	ErrExpectByteSlicePtr = errors.New("expected byte slice reference / pointer argument")
+
+	// ErrExpectByteSlice denotes that the assertion of a byte slice failed
+	ErrExpectByteSlice = errors.New("expected byte slice argument")
+)
+
+// byteDecoder reads bytes from a Reader
+type byteDecoder struct {
+	io.Reader
+	zeroCopy bool
+}
+
+// Decode reads bytes from a Reader
+func (bd *byteDecoder) Decode(v any) error {
+	slice, ok := v.(*[]byte)
+	if !ok {
+		return ErrExpectByteSlicePtr
+	}
+
+	out, err := io.ReadAll(bd)
+	if err != nil {
+		return err
+	}
+
+	if bd.zeroCopy {
+		*slice = out
+	} else {
+		if len(*slice) < len(out) {
+			*slice = make([]byte, len(out))
+		}
+		copy(*slice, out)
+	}
+
+	return nil
+}
+
+// byteEncoder wrties bytes to a Writer
+type byteEncoder struct {
+	io.Writer
+}
+
+// Encode sends bytes down a writer
+func (be *byteEncoder) Encode(v any) error {
+	slice, ok := v.([]byte)
```

Comment:
Above you check for a slice pointer. Is there a reason you don't here?

Reply:
Absolutely, because we don't use a pointer here (because we _read_ from the slice, and like all other `Encoders` that operation is on a non-pointer variable). So this aligns the interface with all the other encoders (which allows using them all in the same manner)...